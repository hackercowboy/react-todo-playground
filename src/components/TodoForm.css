form {
  display: block;
  width: 100%;
  box-sizing: border-box;
}

.todo-form {
  padding: 20px 20px 10px;
}

input[type="text"] {
  width: 100%;
  height: 40px;
  box-sizing: border-box;
  border: 1px solid #ccc;
  border-radius: 0;
  font-size: 20px;
  line-height: 20px;
  font-family: "IBM Plex Sans Thai Looped", sans-serif;
  padding: 10px 10px 5px;
}
