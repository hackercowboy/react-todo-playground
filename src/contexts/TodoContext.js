import React, {
  createContext, useState, useContext, useMemo,
} from 'react';
import PropTypes from 'prop-types';
import { v4 as uuid } from 'uuid';

const TodoContext = createContext();

function TodoContextProvider({ children }) {
  const [todos, setTodos] = useState([
    {
      id: uuid(), created: new Date(), text: 'Learn React', status: 'pending',
    },
    {
      id: uuid(), created: new Date(), text: 'Bring out the trash', status: 'pending',
    },
    {
      id: uuid(), created: new Date(), text: 'Buy grocery', status: 'pending',
    },
    {
      id: uuid(), created: new Date(), text: 'Mow the lawn', status: 'done',
    },
  ]);

  const context = useMemo(() => {
    const addTodo = (todo) => {
      setTodos([...todos, {
        ...todo, id: uuid(), created: new Date(), status: 'pending',
      }]);
    };

    const deleteTodo = (todo) => {
      setTodos(todos.filter((t) => t.id !== todo.id));
    };

    const finishTodo = (todo) => {
      setTodos(todos.map((t) => (t.id === todo.id ? { ...todo, status: 'done' } : t)));
    };
    return {
      todos, addTodo, deleteTodo, finishTodo,
    };
  }, [todos]);

  return (
    <TodoContext.Provider value={context}>
      {children}
    </TodoContext.Provider>
  );
}

const useTodoContext = () => useContext(TodoContext);

TodoContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export { TodoContext, TodoContextProvider, useTodoContext };
