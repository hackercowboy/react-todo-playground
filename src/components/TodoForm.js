import React, { useEffect } from 'react';
import { Formik, Field, Form } from 'formik';

import { useTodoContext } from '../contexts/TodoContext';

import './TodoForm.css';

function TodoForm() {
  const input = React.useRef();
  const { addTodo } = useTodoContext();

  const handleOnSubmit = (values, actions) => {
    try {
      addTodo(values);
      actions.resetForm({ text: '' });
    } finally {
      input.current.focus();
    }
  };

  useEffect(() => {
    input.current.focus();
  });

  return (
    <Formik initialValues={{ text: '' }} onSubmit={handleOnSubmit} role="form">
      <Form className="todo-form" data-testid="todo-form">
        <Field
          id="text"
          name="text"
          type="text"
          placeholder="Add new Todo"
          innerRef={input}
          data-testid="todo-form-input"
        />
      </Form>
    </Formik>
  );
}

export default TodoForm;
