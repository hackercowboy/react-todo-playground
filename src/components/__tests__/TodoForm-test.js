import React from 'react';
import {
  render, prettyDOM, screen, fireEvent, act,
} from '@testing-library/react';

import TodoForm from '../TodoForm';

import { useTodoContext } from '../../contexts/TodoContext';

jest.mock('../../contexts/TodoContext', () => ({
  useTodoContext: jest.fn(),
}));

describe('<TodoForm/>', () => {
  it('should render', () => {
    useTodoContext.mockImplementation(() => ({ addTodo: jest.fn() }));

    render(<TodoForm />);
    expect(prettyDOM()).toMatchSnapshot();
  });

  it('should create new todo on submit', async () => {
    const addTodo = jest.fn();
    useTodoContext.mockImplementation(() => ({ addTodo }));

    render(<TodoForm />);

    const input = screen.getByTestId('todo-form-input');

    await act(async () => {
      await fireEvent.change(input, { target: { value: 'Test' } });
      await fireEvent.submit(input);
    });

    expect(addTodo).toBeCalledWith({ text: 'Test' });
  });
});
