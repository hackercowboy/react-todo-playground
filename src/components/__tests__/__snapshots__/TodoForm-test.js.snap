// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoForm/> should render 1`] = `
"[36m<body>[39m
  [36m<div>[39m
    [36m<form[39m
      [33maction[39m=[32m\\"#\\"[39m
      [33mclass[39m=[32m\\"todo-form\\"[39m
      [33mdata-testid[39m=[32m\\"todo-form\\"[39m
    [36m>[39m
      [36m<input[39m
        [33mdata-testid[39m=[32m\\"todo-form-input\\"[39m
        [33mid[39m=[32m\\"text\\"[39m
        [33mname[39m=[32m\\"text\\"[39m
        [33mplaceholder[39m=[32m\\"Add new Todo\\"[39m
        [33mtype[39m=[32m\\"text\\"[39m
        [33mvalue[39m=[32m\\"\\"[39m
      [36m/>[39m
    [36m</form>[39m
  [36m</div>[39m
[36m</body>[39m"
`;
